### PR TITLE
Immutable cursor

### PR DESCRIPTION
**Pull request by pakrym**
Say not to mutable structures.

### CODE REVIEW

**Comment by benaadams** [comment]
Careful with readonly fields in structs; can cause the struct to duplicate to make a method or property call:

https://codeblog.jonskeet.uk/2014/07/16/micro-optimization-the-surprising-inefficiency-of-readonly-fields/

(Or readonly structs in classes that you call methods or properties on those structs)

**Comment by pakrym** [comment]
Removed readonly, rebased, fixed tests.

**Comment by davidfowl** [comment]
I think I like this change

File: src/Channels/ReadCursor.cs [hunk 1]
```diff
@@ -139,20 +145,19 @@ internal int Seek(int bytes)
...
-                    return bytes;
+                    bytesSeeked = bytes;
+                    return new ReadCursor(segment, index + bytes);
                 }
             }
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal bool TryGetBuffer(ReadCursor end, out PooledBufferSpan span)
+        internal ReadCursor TryGetBuffer(ReadCursor end, out PooledBufferSpan span)
```
Comment (davidfowl) [inline src/Channels/ReadCursor.cs]: This is weird. TryGetBuffer returning a cursor? Should be:

`internal bool TryGetBuffer(ReadCursor end, out PooledBufferSpan span, out ReadCursor cursor)`

File: src/Channels/ReadCursor.cs [hunk 2]
```diff
@@ -165,21 +170,19 @@ internal bool TryGetBuffer(ReadCursor end, out PooledBufferSpan span)
...
             }
             else
             {
-                return TryGetBufferMultiSegment(end, out span);
+                return TryGetBufferMultiBlock(end, out span);
             }
         }
 
-        private bool TryGetBufferMultiSegment(ReadCursor end, out PooledBufferSpan span)
+        private ReadCursor TryGetBufferMultiBlock(ReadCursor end, out PooledBufferSpan span)
```
Comment (davidfowl) [inline src/Channels/ReadCursor.cs]: Same comment as above

File: src/Channels/ReadableBuffer.cs [hunk 3]
```diff
@@ -67,8 +67,7 @@ internal ReadableBuffer(Channel channel, ReadCursor start, ReadCursor end, bool
             _end = end;
             _isOwner = isOwner;
 
-            var begin = start;
-            begin.TryGetBuffer(end, out _span);
+            start.TryGetBuffer(end, out _span, out start);
```
Comment (davidfowl) [inline src/Channels/ReadableBuffer.cs:70]: This look so bizzare 😄

File: src/Channels/ReadableBuffer.cs [hunk 4]
```diff
@@ -246,14 +243,8 @@ public bool TrySliceTo(byte b1, out ReadableBuffer slice, out ReadCursor cursor)
...
         {
-            var begin = _start;
-            if (start != 0)
-            {
-                begin.Seek(start);
-            }
-            var end = begin;
-            end.Seek(length);
-            return Slice(begin, end);
+            var begin = _start.Seek(start);
```
Comment (davidfowl) [inline src/Channels/ReadableBuffer.cs:246]: Why remove the 0 check

**Pull request merged**